### DISCUSSION TITLE
trouble matching tautomeric business rule smarts

### DISCUSSION THREAD

**Discussion by ljmartin** [rdkit/rdkit#7154] [Q&A]
Hi all, 
I'd like to apply syngenta's business rules for tautomer standardization, but am failing to match a query to a molecule that was designed to match it. 

In short, this pattern:
```
patt = Chem.MolFromSmarts('[#6]1(=[#7])-[#6](=[#7])-[#7](-[#1])-[#6]=,:[#7]-[#7]-1-[#1]')
```
isn't a substructure of this molecule, although all the bonds and atoms are correct (as far as I can see! see below for pic):
```
mol = Chem.MolFromSmiles('N=C1NC=NNC1=N')
print('has match?', mol.HasSubstructMatch(patt))
```
```
has match? False
```
I tried adding explicit hydrogens to the molecule, since two hydrogens are explicitly defined in the SMARTS pattern, but no dice: 
```
mol = Chem.MolFromSmiles('N=C1[NH]C=N[NH]C1=N')
print('has match?', mol.HasSubstructMatch(patt))
```
```
has match? False
```

Does anyone understand what I've done wrong? Thank you!

See [gist](https://gist.github.com/ljmartin/d1d51426295e3593d3c93ca3b6bdd879) or pictures here: 
```
Draw.MolsToGridImage(
    [
        Chem.MolFromSmarts('[#6]1(=[#7])-[#6](=[#7])-[#7](-[#1])-[#6]=,:[#7]-[#7]-1-[#1]'),
        Chem.MolFromSmiles('N=C1NC=NNC1=N'),
        Chem.MolFromSmiles('N=C1[NH]C=N[NH]C1=N'),
    ],
    subImgSize=(250,250),
    legends=['Patt','mol', 'molH']
)
```
<img width="733" alt="Screenshot 2024-02-09 at 11 34 06 am" src="https://github.com/rdkit/rdkit/assets/8262765/7bea2f75-59c3-4ac0-a366-a5211c8f627b">

**Comment by ljmartin** [comment]
curiously if I 'port' the SMARTS directly to SMILES (asserting a double bond in place of `=,:`), it still doesn't match:
```
patt = Chem.MolFromSmarts('[#6]1(=[#7])-[#6](=[#7])-[#7](-[#1])-[#6]=,:[#7]-[#7]-1-[#1]')

mol = Chem.MolFromSmiles('C1(=N)-C(=N)-[NH]-C=N-[NH]-1')
mol.HasSubstructMatch(patt)
```
False

**Comment by ljmartin** [comment]
OK this was to do with aromaticity AND explicit hydrogens. Firstly, removing the explicit hydrogens by hand + kekulizing with removal of aromaticity flags works:
```
patt = Chem.MolFromSmarts('[#6]1(=[#7])-[#6](=[#7])-[#7]-[#6]=,:[#7]-[#7]-1') #removed Hs by hand. 
mol = Chem.MolFromSmiles('N=C1NC=NNC1=N')
Chem.Kekulize(mol, clearAromaticFlags=True)
print('has match?', mol.HasSubstructMatch(patt))
```
but this requires editing by hand. `Chem.RemoveHs` does not work on this pattern when explicit Hs are present:
```
patt = Chem.MolFromSmarts('[#6]1(=[#7])-[#6](=[#7])-[#7](-[#1])-[#6]=,:[#7]-[#7]-1-[#1]')
patt = Chem.RemoveHs(patt)
Chem.MolToSmarts(patt)
>>> '[#6]1(=[#7])-[#6](=[#7])-[#7](-[#1])-[#6]=,:[#7]-[#7]-1-[#1]'
```
note Hs are still there.

So I settled on this. Take a round-trip through a RWMol, deleting hydrogens. For each hydrogen deleted, record it's parent atom. Then expand the query on that parent atom to enforce the correct degree. After kekulizing and removing aromaticity flags on the test molecule, it passes. Hopefully this works on a large scale automatically:
```
patt = Chem.MolFromSmarts('[#6]1(=[#7])-[#6](=[#7])-[#7](-[#1])-[#6]=,:[#7]-[#7]-1-[#1]')

mw = Chem.RWMol(patt)
mw.BeginBatchEdit()

for atom in mw.GetAtoms():
    if atom.GetAtomicNum()==1:
        idx = atom.GetIdx()
        bond = [i for i in atom.GetBonds()][0] #hydrogens only have one.
        other_atom =bond.GetOtherAtom(atom)
        other_idx = other_atom.GetIdx()
        #existing degree is going to be one less after we remove this H, 
        #so enforce degree = existing_degree-1
        other_atom.ExpandQuery(rdqueries.ExplicitDegreeEqualsQueryAtom(other_atom.GetDegree()-1))
        mw.RemoveAtom(atom.GetIdx())
mw.CommitBatchEdit()
mol = Chem.MolFromSmiles('N=C1NC=NNC1=N')
print('has match?', mol.HasSubstructMatch(mw))
Chem.Kekulize(mol, clearAromaticFlags=True)
print('kekulize and remove arom flags - has match?', mol.HasSubstructMatch(mw))
```

output:
```
has match? False
kekulize and remove arom flags - has match? True

```

**Comment by DavidACosgrove** [comment]
I think your key problem here is that the bond between atoms 6 and 7 in the pattern is explicitly double or aromatic but all the other ring bonds are explicitly single.  The ring in your test molecule is aromatic.   To be able to stand a chance of matching your molecule, the pattern needs to be ```'[#6]1(=[#7])[#6](=[#7])[#7]([#1])[#6]=,:[#7][#7]1[#1]'```, so the other ring bonds will match 'single or aromatic' which is the default for SMARTS.
![image](https://github.com/rdkit/rdkit/assets/9198870/aff3f6fc-254e-43dc-9110-ba50745a20c7)

The other issue is indeed the explicit H atoms.  An easier way of dealing with this is:
```
mol = Chem.AddHs(mol)
```
Which gives
![image](https://github.com/rdkit/rdkit/assets/9198870/62e98c9c-49b2-4f0e-9f5a-602bb07758f3)
![image](https://github.com/rdkit/rdkit/assets/9198870/f53a72cc-a59e-4884-a447-e2056f69e807)

**Comment by ljmartin** [comment]
hi David - thank you so much lot for looking at this!

Do you know if the reverse is possible - 'aromatizing' the SMARTS pattern? The context is that I'm looking to apply a query/rxn RDFile containing queries and replacement structures (for standardizing tautomers). I could indeed add hydrogens, remove aromatic flags, and iterate over all kekule forms in a library to check if each form is a match. This is not much more human effort, but a bit more computational time on large libraries. From what I can tell, RDFiles like this (originating from Biovia or chemaxon I think) are not well-suited to interfacing with RDKit mols, unless I'm doing something wrong. 

thanks again
